### ISSUE TITLE
Use a named pipe for clients/viewer communication.

### ISSUE THREAD

**Issue by dvinc** [ENSTABretagneRobotics/VIBES#23]
Implement a better threaded input system (with Qt signals for new messages)
- [ ] Threaded message input with generic interface

Subtasks for Linux/MacOS (with mkfifo) and Windows (CreateNamedPipe) implementation.
- [ ] Linux / MacOS X
- [ ] Windows

**Comment by dvinc** [comment]
I'm working on it. It works on MacOS as of now, should be testing on Linux soon. Code has to be cleaned before I commit.
Closing the Viewer while a client is still connected and writing closes the client (SIGPIPE I think). Maybe we should update the API files to handle disconnection more gracefully.

**Comment by SimonRohou** [comment]
In some cases, it could be interesting to keep the JSON file (and even to make it more visible). For instance, VIBES could be used onboard a robot without displaying graphics and the mission could be replayed latter thanks to the JSON file. The company RTSys is currently using it to check the validity of the process.
The user might have the choice between a named pipe or a visible JSON file.

**Comment by dvinc** [comment]
Hi Simon, great to hear that people are using VIBes :smiley: 
In my view (and current ongoing implementation), only the viewer creates the named pipe. The client file opening code is left unchanged.
This implies that:
- if the viewer is not launched, calling `vibes::beginDrawing()` creates a standard file `~/.vibes.json`, not a pipe, so that contents can be retrieved.
- if you explicitly specify a filename when starting the VIBes client `vibes::beginDrawing("filename.json")`, all commands are written to the chosen file (thus addressing your "visible" file concern). If the file does not exists, it is created as a standard file. If the file already exists, VIBes commands data are appended at the end of the file (you need to explicitly remove or clear the file befe starting VIBes if you want to).

For the replay thing, Vibes viewer has a `--show-open-dlg` command line option, that shows a file input dialog at startup to choose a JSON file. JSON files opened this way are not destroyed on exit (contrarily to the default JSON file). We should add an "Open file" button in the Viewer (I created an issue for this).

**Comment by nicolaje** [comment]
@dvinc should we still keep this feature as a requirement for the first non-beta public release?
The textfile approach has proven to work well for a lot of people for the past 2 years. Introducing this now without the 2 years feedback sounds risky.
Shouldn't we mark it for the second public release?

**Comment by dvinc** [comment]
@nicolaje I agree with your idea to retarget this improvement to the second release. My main concern with the textfile approach was the risk of generating huge files on disk (like with the pong example). The auto deletion feature introduced in 26c3cb8 minimizes the problem by introducing some cleanup. However, the location of .vibes.json file in the user's home folder seems inappropriate (wrt to automatic backup programs for instance). I think it should be in /tmp (or Windows equivalent). Do you remember why the user directory has been chosen? (I cannot remember why)

**Comment by nicolaje** [comment]
Great. Targeting the home folder looked like a good universal approach at the time, but indeed, the TMP (%TMP% or %TEMP% on Windows) is more appropriate as it won't seem to interfere with the user's home directory.
Let's change the directory for the 1st public release.
